.github/workflows/test_with_clone.yaml

```diff
@@ -0,0 +1,179 @@
+
+            - name: Sanitize Clone Name
+              run: |
+                CLONE_NAME_SANITIZED="${CLONE_NAME//[^a-zA-Z0-9_]/_}"
+                echo "Clone name has been updated! The clone name will be: ${CLONE_NAME_SANITIZED}"
+                echo "SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE=${CLONE_NAME_SANITIZED}" >> $GITHUB_ENV
+
+            - name: Zero-copy clone the database
+              shell: bash
+              run: |
+                snow sql -q "CREATE OR REPLACE DATABASE $SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE CLONE $SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE_ORIG;"
+
+            - name: Grant permissions to DATA_ENGINEER on cloned database
+              shell: bash
+              run: |
+                snow connection set-default sysadmin
+
+                # Transfer ownership of: database
+                snow sql -q "GRANT OWNERSHIP ON DATABASE ${SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE} TO ROLE DATA_ENGINEER REVOKE CURRENT GRANTS;"
+
+                # Transfer ownership of: schemas
+                snow sql -q "GRANT OWNERSHIP ON SCHEMA ${SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE}.SYNAPSE_RAW TO ROLE DATA_ENGINEER REVOKE CURRENT GRANTS;"
+                snow sql -q "GRANT OWNERSHIP ON SCHEMA ${SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE}.SYNAPSE TO ROLE DATA_ENGINEER REVOKE CURRENT GRANTS;"
+                
+                # Transfer ownership of: tables
+                snow sql -q "GRANT OWNERSHIP ON ALL TABLES IN SCHEMA ${SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE}.SYNAPSE_RAW TO ROLE DATA_ENGINEER REVOKE CURRENT GRANTS;"
+                snow sql -q "GRANT OWNERSHIP ON ALL TABLES IN SCHEMA ${SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE}.SYNAPSE TO ROLE DATA_ENGINEER REVOKE CURRENT GRANTS;"
+                
+                # Transfer ownership of: dynamic tables
+                snow sql -q "GRANT OWNERSHIP ON ALL DYNAMIC TABLES IN SCHEMA ${SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE}.SYNAPSE TO ROLE DATA_ENGINEER REVOKE CURRENT GRANTS;"
+
+            - name: Run schemachange on the clone as DATA_ENGINEER
+              shell: bash
+              run: |
+                schemachange \
+                -f synapse_data_warehouse \
+                -a $SNOWFLAKE_ACCOUNT \
+                -u $SNOWFLAKE_USER \
+                -r $SNOWFLAKE_CLONE_ROLE \
+                -w $SNOWFLAKE_WAREHOUSE \
+                --config-folder synapse_data_warehouse
+
+    drop_clone:
+      runs-on: ubuntu-latest
+      if: github.event.pull_request.merged == true || github.event.action == 'closed'
+      environment: dev
+      env:
+          SNOWFLAKE_PASSWORD: ${{ secrets.SNOWSQL_PWD }}
+          SNOWFLAKE_ACCOUNT: ${{ secrets.SNOWSQL_ACCOUNT }}
+          SNOWFLAKE_USER: ${{ secrets.SNOWSQL_USER }}
+          SNOWFLAKE_WAREHOUSE: ${{ secrets.SNOWSQL_WAREHOUSE }}
+          SNOWFLAKE_CLONE_ROLE: DATA_ENGINEER
+          CLONE_NAME: "${{ vars.SNOWFLAKE_SYNAPSE_DATA_WAREHOUSE_DATABASE }}_${{ github.head_ref }}"
```

Comment:
If I am understanding this will only drop the clone when the PR is merged or closed. However, a cloned DB is created on any push when the PR is labeled and open.

Would this mean that only the last `github.head_ref` is cleaned up? In the end we would want to cleanup both the `head_ref` when things are merged, AND, when a new push is added (and head_ref is updated) we'd also want to drop the previous cloned DB.

Reply:
> Would this mean that only the last github.head_ref is cleaned up?

Good question - Yes, this means that the  last `github.head_ref` is cleaned up. However, this is not a concern because it's the _only_ cloned DB on snowflake, for a given feature branch. All predecessor clones just get replaced by a new clone.

So for example if `head_ref` is `snow-90-my-branch`, the procedure goes like this:

1. an initial commit is made + job is triggered
2. new clone is made named: `synapse_data_warehouse_dev_snow_90_my_branch`
3. schemachange runs
4. another commit is made
5. **old clone is replaced by the new clone of `synapse_data_warehouse_dev` with a `CREATE OR REPLACE`. the new clone has the same name as the last clone, so it does a `REPLACE`**
6. schemachange runs
7. rinse and repeat
8. when it's time to merge, the clone is dropped (if it still exists). the drop clone job assumes the name hasn't been changed since its initial create, so this is a potential edge case actually, but it's probably enough to just say "don't change the clone name" in the CONTRIBUTIONS doc